casper/src/test/scala/coop/rchain/casper/addblock/MultiParentCasperAddBlockSpec.scala

```diff
@@ -558,4 +558,51 @@ class MultiParentCasperAddBlockSpec extends FlatSpec with Matchers with Inspecto
       } yield block
     }
   }
+
+  it should "merge dag containing merge blocks, if only one block per validator is in scope" in effectTest {
+    TestNode.networkEff(genesis, networkSize = 5).use {
+      case nodes @ n0 +: n1 +: n2 +: n3 +: n4 +: _ =>
+        for {
+          deploys <- (genesis.validatorPks zip genesis.genesisVaultsSks).zipWithIndex.toList
+                      .traverse {
+                        case ((_, payerSk), i) => {
+                          ConstructDeploy.basicDeployData[Effect](i, payerSk)
+                        }
+                      }
+          //b0
+          b0 <- n0.createBlockUnsafe(deploys.head)
+          _  <- nodes.toList.traverse(_.processBlock(b0))
+          _  = b0.header.parentsHashList.size shouldBe 1
+          //   b1 b2
+          //b0
+          b1 <- n1.createBlockUnsafe(deploys(1))
+          b2 <- n2.createBlockUnsafe(deploys(2))
+          _  <- nodes.toList.traverse(_.processBlock(b1))
+          _  <- nodes.toList.traverse(_.processBlock(b2))
+          _  = b1.header.parentsHashList.size shouldBe 1
+          _  = b2.header.parentsHashList.size shouldBe 1
+          //         b3 b4
+          //   b1 b2
+          //b0
+          b3 <- n3.createBlockUnsafe(deploys(3))
+          b4 <- n4.createBlockUnsafe(deploys(4))
+          _  = (b3.body.state.preStateHash == b4.body.state.preStateHash) shouldBe true
+          _  <- nodes.toList.traverse(_.processBlock(b3))
+          // after b3 processed by all blocks, b4 processing should not be influenced by b3 (b3 should not be used as parent)
```

Comment:
this particular line is about per block view. I'm thinking how to write better tests, but having KeyValueDagRepresentation private class of BlockDagKeyValueStorage makes things not easy. Does it makes sense to just make it public? @tgrospic

Reply:
Makes sense, but I'm not sure that public access will be enough.